Add Bermuda holidays

kritibirda opened the pull request:
Resolves #2412.

<!--
  Thanks for contributing to holidays!
-->

## Proposed change

<!--
  Describe the big picture of your changes.
  Don't forget to link your PR to an existing issue if any.
-->

Add Bermuda holidays

## Type of change

<!--
  Type of change you want to introduce. Please, check one (1) box only!
  If your PR requires multiple boxes to be checked, most likely it needs to
  be split into multiple PRs.
-->

- [X] New country/market holidays support (thank you!)
- [ ] Supported country/market holidays update (calendar discrepancy fix, localization)
- [ ] Existing code/documentation/test/process quality improvement (best practice, cleanup, refactoring, optimization)
- [ ] Dependency update (version deprecation/pin/upgrade)
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] Breaking change (a code change causing existing functionality to break)
- [ ] New feature (new `holidays` functionality in general)

## Checklist

<!--
  Put an `x` in the boxes that apply. You can change them after PR is created.
-->

- [X] I've followed the [contributing guidelines][contributing-guidelines]
- [X] I've successfully run `make check`, all checks and tests are green

<!--
  Thanks again for your contribution!
-->

[contributing-guidelines]: https://github.com/vacanza/holidays/blob/dev/CONTRIBUTING.md
[docs]: https://github.com/vacanza/holidays/tree/dev/docs/source

kritibirda committed:
Add Bermuda holidays

Resolves #2412.

kritibirda26 commented:
I could not find the original text of the Public Holidays Amendment Act 1947, only the consolidated version containing all the amendments. But I have implemented the holidays by trying to reconstruct it from the amendments and old captures from Internet Archive.

KJhellico commented on holidays/countries/bermuda.py:
```diff
@@ -0,0 +1,136 @@
+        if dt is not None:
+            self._add_observed(dt)
+
+        if self._year <= 1999:
+            # Cup Match Day.
+            name = "Cup Match Day"
+        else:
+            # Emancipation Day.
+            name = "Emancipation Day"
+        self._add_observed(self._add_holiday_4_days_prior_1st_mon_of_aug(name))
+
+        name = (
+            # Somers Day.
+            "Somers Day"
+            if self._year <= 2020
+            # Mary Prince Day.
+            else "Mary Prince Day"
+        )
+        self._add_observed(self._add_holiday_3_days_prior_1st_mon_of_aug(name))
```
```suggestion
        self._add_holiday_4_days_prior_1st_mon_of_aug(
            # Cup Match Day.
            "Cup Match Day"
            if self._year <= 1999
            # Emancipation Day.
            else "Emancipation Day"
        )

        self._add_holiday_3_days_prior_1st_mon_of_aug(
            # Somers Day.
            "Somers Day"
            if self._year <= 2020
            # Mary Prince Day.
            else "Mary Prince Day"
        )
```

KJhellico commented on holidays/countries/bermuda.py:
```diff
@@ -0,0 +1,136 @@
+        self._add_observed(self._add_holiday_3_days_prior_1st_mon_of_aug(name))
+
+        # Labour Day.
+        self._add_observed(self._add_holiday_1st_mon_of_sep("Labour Day"))
```
```suggestion
        self._add_holiday_1st_mon_of_sep("Labour Day")
```

KJhellico commented on tests/countries/test_bermuda.py:
```diff
@@ -0,0 +1,187 @@
+        name = "New Year's Day"
+        self.assertHolidayName(name, (f"{year}-01-01" for year in range(1948, 2050)))
+        dt = (
+            "1995-01-02",
+            "2006-01-02",
+            "2012-01-02",
+            "2017-01-02",
+            "2023-01-02",
```
```suggestion
            "2011-01-03",
            "2012-01-02",
            "2017-01-02",
            "2022-01-03",
            "2023-01-02",
```

KJhellico commented on tests/countries/test_bermuda.py:
```diff
@@ -0,0 +1,187 @@
+
+    def test_bermuda_day(self):
+        name = "Bermuda Day"
+        for year in range(1948, 2018):
+            self.assertHolidayName(
+                name,
+                f"{year}-05-24",
+                (
+                    "2018-05-25",
+                    "2019-05-31",
+                    "2020-05-22",
+                    "2021-05-28",
+                    "2022-05-27",
+                    "2023-05-26",
+                    "2024-05-24",
+                    "2025-05-23",
+                ),
+            )
```
```suggestion
        self.assertHolidayName(name, (f"{year}-05-24" for year in range(1948, 2018)))
        self.assertHolidayName(
            name,
            "2018-05-25",
            "2019-05-31",
            "2020-05-22",
            "2021-05-28",
            "2022-05-27",
            "2023-05-26",
            "2024-05-24",
            "2025-05-23",
        )
        self.assertHolidayName(name, range(2018, 2050))
```

KJhellico commented on tests/countries/test_bermuda.py:
```diff
@@ -0,0 +1,187 @@
+        self.assertNoNonObservedHoliday(dt)
+
+    def test_queens_birthday(self):
+        name = "Queen's Birthday"
```
```suggestion
        name = "Queen's Birthday"
        self.assertHolidayName(
            name,
            "2003-06-11",
            "2004-06-16",
            "2005-06-15",
            "2006-06-14",
            "2007-06-13",
            "2008-06-11",
        )
```

KJhellico commented on tests/countries/test_bermuda.py:
```diff
@@ -0,0 +1,187 @@
+
+    def test_national_heroes_day(self):
+        name = "National Heroes Day"
+        self.assertNoHolidayName(name, range(1948, 2008))
+        self.assertHolidayName(
+            name,
+            (
+                "2008-10-13",
+                "2009-06-15",
+                "2010-06-21",
+                "2020-06-15",
+                "2021-06-21",
+                "2022-06-20",
+                "2023-06-19",
+                "2024-06-17",
+                "2025-06-16",
+            ),
+        )
```
```suggestion
        self.assertHolidayName(
            name,
            "2008-10-13",
            "2009-06-15",
            "2010-06-21",
            "2020-06-15",
            "2021-06-21",
            "2022-06-20",
            "2023-06-19",
            "2024-06-17",
            "2025-06-16",
        )
        self.assertHolidayName(name, range(2008, 2050))
        self.assertNoHolidayName(name, range(1948, 2008))
```

KJhellico commented on tests/countries/test_bermuda.py:
```diff
@@ -0,0 +1,187 @@
+        )
+
+    def test_somers_and_mary_prince_day(self):
+        name_1 = "Somers Day"
```
```suggestion
        name_1 = "Somers Day"
        self.assertHolidayName(
            name_1,
            "2016-07-29",
            "2017-08-04",
            "2018-08-03",
            "2019-08-02",
            "2020-07-31",
        )
```

KJhellico commented on tests/countries/test_bermuda.py:
```diff
@@ -0,0 +1,187 @@
+        self.assertNoHolidayName(name_1, range(2021, 2050))
+
+        name_2 = "Mary Prince Day"
+        self.assertNoHolidayName(name_2, range(1948, 2021))
+        self.assertHolidayName(name_2, range(2021, 2050))
```
```suggestion
        self.assertHolidayName(
            name_2,
            "2021-07-30",
            "2022-07-29",
            "2023-08-04",
            "2024-08-02",
            "2025-08-01",
        )
        self.assertHolidayName(name_2, range(2021, 2050))
        self.assertNoHolidayName(name_2, range(1948, 2021))
```

KJhellico commented on tests/countries/test_bermuda.py:
```diff
@@ -0,0 +1,187 @@
+                "2025-06-16",
+            ),
+        )
+
```
Cup Match / Emancipation Day test is lost.

KJhellico commented on tests/countries/test_bermuda.py:
```diff
@@ -0,0 +1,187 @@
+
+    def test_labour_day(self):
+        name = "Labour Day"
+        self.assertHolidayName(name, range(1948, 2050))
+        self.assertHolidayName(
+            name,
+            "2020-09-07",
+            "2021-09-06",
+            "2022-09-05",
+            "2023-09-04",
+            "2024-09-02",
+            "2025-09-01",
+        )
```
```suggestion
        self.assertHolidayName(
            name,
            "2020-09-07",
            "2021-09-06",
            "2022-09-05",
            "2023-09-04",
            "2024-09-02",
            "2025-09-01",
        )
        self.assertHolidayName(name, range(1948, 2050))
```

KJhellico commented on tests/countries/test_bermuda.py:
```diff
@@ -0,0 +1,187 @@
+        self.assertHolidayName(name, (f"{year}-11-11" for year in range(1948, 2050)))
+
+        dt = (
+            "2001-11-12",
+            "2007-11-12",
+            "2012-11-12",
+            "2018-11-12",
+            "2029-11-12",
```
```suggestion
            "2007-11-12",
            "2012-11-12",
            "2017-11-13",
            "2018-11-12",
            "2023-11-13",
```

KJhellico commented on tests/countries/test_bermuda.py:
```diff
@@ -0,0 +1,187 @@
+        name = "Christmas Day"
+        self.assertHolidayName(name, (f"{year}-12-25" for year in range(1948, 2050)))
+        dt = (
+            "2004-12-27",
+            "2005-12-27",
+            "2010-12-27",
+            "2011-12-27",
+            "2016-12-27",
+            "2022-12-27",
```
```suggestion
            "2010-12-27",
            "2011-12-27",
            "2016-12-27",
            "2021-12-27",
            "2022-12-27",
```

KJhellico commented on tests/countries/test_bermuda.py:
```diff
@@ -0,0 +1,187 @@
+        name = "Boxing Day"
+        self.assertHolidayName(name, (f"{year}-12-26" for year in range(1948, 2050)))
+        dt = (
+            "2010-12-28",
+            "2021-12-28",
```
```suggestion
            "2009-12-28",
            "2010-12-28",
            "2015-12-28",
            "2020-12-28",
            "2021-12-28",
```

KJhellico reviewed (commented):
I haven't looked at the sources yet, so just technical suggestions.

PPsyrius commented on holidays/countries/bermuda.py:
```diff
@@ -0,0 +1,136 @@
+        * [Public Holidays Amendment Act 2008](https://web.archive.org/web/20250527163810/https://www.bermudalaws.bm/Laws/Annual%20Law/Acts/2008/Public%20Holidays%20Amendment%20Act%202008)
+        * [Public Holidays Amendment (No. 2) Act 2009](https://web.archive.org/web/20250527163816/https://www.bermudalaws.bm/Laws/Annual%20Law/Acts/2009/Public%20Holidays%20Amendment%20%28No.%202%29%20Act%202009)
+        * [Public Holidays Amendment Act 2017](https://web.archive.org/web/20250527163819/https://www.bermudalaws.bm/Laws/Annual%20Law/Acts/2017/Public%20Holidays%20Amendment%20Act%202017)
+        * [Public Holidays Amendment Act 2020](https://web.archive.org/web/20250527163836/https://www.bermudalaws.bm/Laws/Annual%20Law/Acts/2020/Public%20Holidays%20Amendment%20Act%202020)
```
We might want to add this as well as our general cross-checking source:
```suggestion
        * [Public Holidays Amendment Act 2020](https://web.archive.org/web/20250527163836/https://www.bermudalaws.bm/Laws/Annual%20Law/Acts/2020/Public%20Holidays%20Amendment%20Act%202020)
        * [Government of Bermuda](https://web.archive.org/web/20250530055854/https://www.gov.bm/public-holidays)
```

kritibirda committed:
Apply suggestions from code review

Co-authored-by: ~Jhellico <KJhellico@users.noreply.github.com>
Co-authored-by: Panpakorn Siripanich <19505219+PPsyrius@users.noreply.github.com>
Signed-off-by: Kriti Birda <164247895+kritibirda26@users.noreply.github.com>

kritibirda committed:
Merge branch 'dev' into bermuda

kritibirda committed:
update tests

add cup match/emancipation day test
fix somers days/mary prince test
fix bermuda day test for 2020

kritibirda committed:
add special holiday, fix bermuda day 2019 and add sources

KJhellico commented on holidays/countries/bermuda.py:
```diff
@@ -0,0 +1,141 @@
+        # Queen's Birthday.
+        name = "Queen's Birthday"
+        if self._year <= 1999:
+            self._add_holiday_3_days_prior_last_mon_of_may(name)
+        elif self._year <= 2008:
+            self._add_holiday_2_days_past_2nd_mon_of_jun(name)
```
```suggestion
            self._add_holiday_3rd_mon_of_june(name)
        elif self._year <= 2008:
            self._add_holiday_2_days_past_2nd_sat_of_jun(name)
```
Public Holidays Amendment Act 1999:
> in the entry relating to the Queen's Birthday by deleting the words "the third Monday in June" and substituting the words "the Monday following the second Saturday in June";

kritibirda committed:
Update holidays/countries/bermuda.py

Co-authored-by: ~Jhellico <KJhellico@users.noreply.github.com>
Signed-off-by: Kriti Birda <164247895+kritibirda26@users.noreply.github.com>

KJhellico commented on holidays/countries/bermuda.py:
```diff
@@ -0,0 +1,141 @@
+        # Queen's Birthday.
+        name = "Queen's Birthday"
+        if self._year <= 1999:
+            self._add_holiday_3rd_mon_of_june(name)
```
Yes, it's my typo. :(

kritibirda committed:
fix queen's birthday

KJhellico commented:
Already LGTM, but maybe it's worth adding l10n (en_BM, en_US), since Labour Day holiday is present?

arkid15r reviewed (commented):
LGTM, great start for GSoC coding phase!
